pipeline/eval/eval.py

```diff
@@ -72,6 +74,81 @@
     WANDB_AVAILABLE = False
 
 
+def tokenize_nospace(sentence, tokenizer):
+    """
+    Tokenize a sentence and return a list of tokens that have no space
+    """
+
+    def iter_tok():
+        for t in tokenizer.tokenize(sentence):
+            if t == tokenizer.SPACE_TOKEN:
+                continue
+            # simalign removes surrogate and the CI fails because poor models
+            # produce sentences wiht only surrogates and it fails
+            # replace them so alignment can run
+            if t == "�":
+                yield "?"
+            yield t
+
+    return list(iter_tok())
+
+
+def filter_empty(s, t):
+    """
+    Filter out sentence pairs that are empty
+    """
+    for i, j in zip(s, t):
+        if not i or not j:
+            continue
+        yield i, j
+
+
+def compute_unaliged_ratio(src: str, trg: str, source_lines: List[str], target_lines: List[str]):
+    """
+    Compute ratio of unaligned tokens
+    """
+    aligner = SentenceAligner(model="bert", token_type="bpe", matching_methods="mai")
+    src_tokenizer = IcuTokenizer(src)
+    trg_tokenizer = IcuTokenizer(trg)
+    src_tokens = [tokenize_nospace(i, src_tokenizer) for i in source_lines]
+    trg_tokens = [tokenize_nospace(i, trg_tokenizer) for i in target_lines]
+    src_tokens, trg_tokens = zip(*filter_empty(src_tokens, trg_tokens))
+
+    def gen_alignments():
+        for st, tt in zip(src_tokens, trg_tokens):
+            try:
+                yield aligner.get_word_aligns(st, tt)
+            except ValueError as e:
+                # When there is a sentences containing only characters that are not present
+                # in the sentencealigner vocab, it cannot align because there are no embeddings
+                # so, in this case we just return an empty alignment pairs
+                # which means nothing could be aligned
+                if e.args and e.args[0].startswith("Found array with 0 sample"):
+                    yield {"itermax": []}
+                else:
+                    raise e
+            except Exception as e:
+                # If it fails print sentence pair for easier debugging
+                logger.error(f"Source: {st}")
```

Comment:
Should we add an error message before printing to provide a context on what those errors are?